mhcflurry/dataset.py

```diff
@@ -0,0 +1,554 @@
+                affinities.append(affinity)
+        return cls.from_sequences(
+            alleles=alleles,
+            peptides=peptides,
+            affinities=affinities)
+
+    @classmethod
+    def create_empty(cls):
+        """
+        Returns an empty Dataset containing no pMHC entries.
+        """
+        return cls.from_nested_dictionary({})
+
+    @classmethod
+    def from_single_allele_dictionary(
+            cls,
+            allele_name,
+            peptide_to_affinity_dict):
+        """
+        Given a peptide->affinity dictionary for a single allele,
+        create a Dataset.
+        """
+        return cls.from_nested_dictionary({allele_name: peptide_to_affinity_dict})
+
+    @classmethod
+    def from_csv(
+            cls,
+            filename,
+            sep=None,
+            allele_column_name=None,
+            peptide_column_name=None,
+            affinity_column_name=None):
+        df, allele_column_name, peptide_column_name, affinity_column_name = \
+            load_dataframe(
+                filename=filename,
+                sep=sep,
+                allele_column_name=allele_column_name,
+                peptide_column_name=peptide_column_name,
+                affinity_column_name=affinity_column_name)
+        df = df.rename(columns={
+            allele_column_name: "allele",
+            peptide_column_name: "peptide",
+            affinity_column_name: "affinity"})
+        return cls(df)
+
+    def get_allele(self, allele_name):
+        """
+        Get Dataset for a single allele
+        """
+        if allele_name not in self.unique_alleles():
+            raise KeyError("Allele '%s' not found" % (allele_name,))
+        df = self.to_dataframe()
+        df_allele = df[df.allele == allele_name]
+        return self.__class__(df_allele)
+
+    def get_alleles(self, allele_names):
+        """
+        Restrict Dataset to several allele names.
+        """
+        datasets = []
+        for allele_name in allele_names:
+            datasets.append(self.get_allele(allele_name))
+        return self.concat(datasets)
+
+    @classmethod
+    def concat(cls, datasets):
+        """
+        Concatenate several datasets into a single object.
+        """
+        dataframes = [dataset.to_dataframe() for dataset in datasets]
+        return cls(pd.concat(dataframes))
+
+    def replace_allele(self, allele_name, new_dataset):
+        """
+        Replace data for given allele with new entries.
+        """
+        if allele_name not in self.unique_alleles():
+            raise ValueError("Allele '%s' not found" % (allele_name,))
+        df = self.to_dataframe()
+        df_without = df[df.allele != allele_name]
+        new_df = new_dataset.to_dataframe()
+        combined_df = pd.concat([df_without, new_df])
+        return self.__class__(combined_df)
+
+    def flatmap_peptides(self, peptide_fn):
+        """
+        Create zero or more peptides from each pMHC entry. The affinity of all
```

Comment:
I think documenting that `peptide_fn` is of type `string -> string list` (is that right?) would be useful